op-program/client/l2/engine_backend.go

```diff
@@ -218,29 +198,7 @@ func (o *OracleBackedL2Chain) AssembleAndInsertBlockWithoutSetHead(processor *en
 }
 
 func (o *OracleBackedL2Chain) SetCanonical(head *types.Block) (common.Hash, error) {
-	oldHead := o.head
```

Comment:
Moving this code over is good reuse but the slight changes worry me.

Reply:
The only change made is to short-circuit the walk-back by tracking the earliest header seen. The old logic didn't do this. (Which wasn't a bug per se, but made walk-back accidentally quadratic).

Should be fixed now.